Add support for float and double precision data

mlxd opened the pull request:
**Context:** This PR will enable support for 64bit and 128bit complex datatypes, as well as restructuring the package class to allow for this extension.

**Description of the Change:** To enable the above, several modifications were required:
- All gates kernels previously defined in `Gates.hpp`and `Gates.cpp` are now part of the `StateVector` class, with gate numeric values accessible in `Gates.hpp`.
- The apply functionality from `Apply.hpp` and `Apply.cpp` has been added to the `StateVector` class directly
- The `StateVector` class was converted to a templated class, allowing all above functions to be compiled with the correct data types.
- The C++ version was bumped from 11 to 17, to allow for use of newer features (`inline static const`, template argument deduction, constexpr if)
- The `StateVector` class and methods can now be directly accessed from Python, using the `StateVectorC64` and `StateVectorC128` types. Use will require adaption of core Python implementation to utilise.
- Tests based on the previous architecture format were removed and updated with Catch2.

**Benefits:** This enables use of data-types previously unavailable without casting (np.complex64), and will allow for memory and performance savings when used. Features spread over multiple additional classes and files are now made redundant, allowing for a more straight-forward implementation.

**Possible Drawbacks:** Structure of `StateVector` class has increased in size.

**Related GitHub Issues:**

mlxd commented:
Note: due to the required refactor of the testing infrastructure, the PR #115 should accompany this review, and will be merged in here before any merge with master.

trbromley commented:
[ch7526] [ch7552] [ch7067]

trbromley commented on pennylane_lightning/src/Util.hpp:
```diff
@@ -18,19 +18,28 @@
+
 /**
- * Calculates 2^n for some integer n > 0 using bitshifts.
+ * Calculates 2^n -1 for some integer n > 0 using bitshifts.
```
Good spot!

trbromley reviewed (approved):
Thanks @mlxd! This looks to have tidied things up nicely :100:

mlxd commented on CMakeLists.txt:
```diff
@@ -20,10 +20,9 @@ set_pennylane_lightning_version("${CMAKE_SOURCE_DIR}/pennylane_lightning/_versio
 set(PROJECT_VERSION ${VERSION_STRING})
 
-set(CMAKE_CXX_STANDARD 11) # At least C++11 is required
+set(CMAKE_CXX_STANDARD 17) # At least C++17 is required
```
Not that I can see. Any system that ships with Python 3.7 natively should have a supported GCC/G++ compiler with c++17 support. It also simplifies some implementation details, so I'd be in favour of it.

mlxd commented on pennylane_lightning/lightning_qubit.py:
```diff
@@ -136,7 +142,7 @@ def apply_lightning(self, state, operations):
 
         state_vector = np.ravel(state)
-        apply(state_vector, op_names, op_wires, op_param, op_inverse, self.num_wires)
+        apply(state_vector, op_names, op_wires, op_inverse, op_param)
```
Yep. With this implementation, the total number of wires can be read from the statevector itself.

mlxd commented on pennylane_lightning/lightning_qubit.py:
```diff
@@ -16,9 +16,9 @@
-from .lightning_qubit_ops import apply
 import numpy as np
 from pennylane import QubitStateVector, BasisState, DeviceError, QubitUnitary
+from .lightning_qubit_ops import apply, StateVectorC64, StateVectorC128
```
Those are the explicitly listed classes for the two statevector precisions. We don't make use of them here. Maybe It is best to remove these for now, and tackle the rewrite here using them a little later.

mlxd commented on pennylane_lightning/src/Bindings.cpp:
```diff
@@ -11,38 +11,105 @@
+        .def(py::init<
+             py::array_t<complex<PrecisionT>,
+                         py::array::c_style | py::array::forcecast> &>())
+        .def("apply", &StateVecBinder<PrecisionT>::apply)
+        .def("PauliX", &StateVecBinder<PrecisionT>::applyPauliX)
+        .def("PauliY", &StateVecBinder<PrecisionT>::applyPauliY)
+        .def("PauliZ", &StateVecBinder<PrecisionT>::applyPauliZ)
+        .def("Hadamard", &StateVecBinder<PrecisionT>::applyHadamard)
+        .def("S", &StateVecBinder<PrecisionT>::applyS)
+        .def("T", &StateVecBinder<PrecisionT>::applyT)
+        .def("CNOT", &StateVecBinder<PrecisionT>::applyCNOT)
+        .def("SWAP", &StateVecBinder<PrecisionT>::applySWAP)
+        .def("CSWAP", &StateVecBinder<PrecisionT>::applyCSWAP)
+        .def("Toffoli", &StateVecBinder<PrecisionT>::applyToffoli)
+        .def("CZ", &StateVecBinder<PrecisionT>::applyCZ)
+        .def("PhaseShift",
+             &StateVecBinder<PrecisionT>::template applyPhaseShift<float>)
+        .def("PhaseShift",
+             &StateVecBinder<PrecisionT>::template applyPhaseShift<double>)
+        .def("RX", &StateVecBinder<PrecisionT>::template applyRX<float>)
+        .def("RX", &StateVecBinder<PrecisionT>::template applyRX<double>)
+        .def("RY", &StateVecBinder<PrecisionT>::template applyRY<float>)
+        .def("RY", &StateVecBinder<PrecisionT>::template applyRY<double>)
+        .def("RZ", &StateVecBinder<PrecisionT>::template applyRZ<float>)
+        .def("RZ", &StateVecBinder<PrecisionT>::template applyRZ<double>)
+        .def("Rot", &StateVecBinder<PrecisionT>::template applyRot<float>)
+        .def("Rot", &StateVecBinder<PrecisionT>::template applyRot<double>)
+        .def("CRX", &StateVecBinder<PrecisionT>::template applyCRX<float>)
+        .def("CRX", &StateVecBinder<PrecisionT>::template applyCRX<double>)
+        .def("CRY", &StateVecBinder<PrecisionT>::template applyCRY<float>)
+        .def("CRY", &StateVecBinder<PrecisionT>::template applyCRY<double>)
+        .def("CRZ", &StateVecBinder<PrecisionT>::template applyCRZ<float>)
+        .def("CRZ", &StateVecBinder<PrecisionT>::template applyCRZ<double>)
+        .def("CRot", &StateVecBinder<PrecisionT>::template applyCRot<float>)
+        .def("CRot", &StateVecBinder<PrecisionT>::template applyCRot<double>);
```
Yep. These can all be called from the `StateVectorCxyz` classes, and will act direct on the data.

mlxd commented on pennylane_lightning/src/Bindings.cpp:
```diff
@@ -11,38 +11,105 @@
+};
+
+template <class PrecisionT> void lightning_class_bindings(py::module &m) {
+    // Enable module name to be based on size of complex datatype
```
Completely agree. Though, I left it a little bare as there was a story to document everything in here (I'd imagine a day or so worth of effort to get it all correct, with docs and all)

mlxd commented on pennylane_lightning/src/StateVector.hpp:
```diff
@@ -15,19 +15,717 @@
-    const size_t length;
+    StateVector()
+        : arr_{nullptr}, length_{0}, num_qubits_{0}, gate_wires_{}, gates_{} {};
+    StateVector(CFP_t *arr, size_t length)
+        : arr_{arr}, length_{length}, num_qubits_{Util::log2(length_)},
+          gate_wires_{
+              {"PauliX", 1}, {"PauliY", 1}, {"PauliZ", 1},     {"Hadamard", 1},
+              {"T", 1},      {"S", 1},      {"RX", 1},         {"RY", 1},
+              {"RZ", 1},     {"Rot", 1},    {"PhaseShift", 1}, {"CNOT", 2},
+              {"SWAP", 2},   {"CZ", 2},     {"CRX", 2},        {"CRY", 2},
+              {"CRZ", 2},    {"CRot", 2},   {"CSWAP", 3},      {"Toffoli", 3}},
```
Sure thing. The previous implementation had some logic to check the gate classes for the number of supported wires to act upon. Since the classes are now methods, the mapping of a gate label to supported wires needed to be elsewhere. I created a private dictionary to map between gate labels and the number of supported wires (`gate_wires_`). This allows minimal changes to the apply methods to match the older functionality. 

Since the dictionary is declared as `const` it must be initialised with this initialiser list syntax. Though, we can move these to a `inline static const / constexpr` and put it outside the class instantiation, though for now I think this is fine.

trbromley replied:
Thanks!

mlxd commented on pennylane_lightning/src/StateVector.hpp:
```diff
@@ -15,19 +15,717 @@
+    }
+    static constexpr vector<CFP_t> getPauliZ() {
+        return {ONE, ZERO, ZERO, -ONE};
+    }
+    static constexpr vector<CFP_t> getHadamard() {
+        return {INVSQRT2, INVSQRT2, INVSQRT2, -INVSQRT2};
+    }
+    static constexpr vector<CFP_t> getS() { return {ONE, ZERO, ZERO, IMAG}; }
+    static constexpr vector<CFP_t> getT() { return {ONE, ZERO, ZERO, IMAG}; }
+    static constexpr vector<CFP_t> getCNOT() {
+        return {ONE,  ZERO, ZERO, ZERO, ZERO, ONE,  ZERO, ZERO,
+                ZERO, ZERO, ZERO, ONE,  ZERO, ZERO, ONE,  ZERO};
+    }
+    static constexpr vector<CFP_t> getSWAP() {
+        return {ONE,  ZERO, ZERO, ZERO, ZERO, ZERO, ONE,  ZERO,
+                ZERO, ONE,  ZERO, ZERO, ZERO, ZERO, ZERO, ONE};
+    }
+    static constexpr vector<CFP_t> getCZ() {
+        return {ONE,  ZERO, ZERO, ZERO, ZERO, ONE,  ZERO, ZERO,
+                ZERO, ZERO, ONE,  ZERO, ZERO, ZERO, ZERO, -ONE};
+    }
+
+    static constexpr vector<CFP_t> getCSWAP() {
+        return {ONE,  ZERO, ZERO, ZERO, ZERO, ZERO, ZERO, ZERO, ZERO, ONE,
+                ZERO, ZERO, ZERO, ZERO, ZERO, ZERO, ZERO, ZERO, ONE,  ZERO,
+                ZERO, ZERO, ZERO, ZERO, ZERO, ZERO, ZERO, ONE,  ZERO, ZERO,
+                ZERO, ZERO, ZERO, ZERO, ZERO, ZERO, ONE,  ZERO, ZERO, ZERO,
+                ZERO, ZERO, ZERO, ZERO, ZERO, ZERO, ONE,  ZERO, ZERO, ZERO,
+                ZERO, ZERO, ZERO, ONE,  ZERO, ZERO, ZERO, ZERO, ZERO, ZERO,
+                ZERO, ZERO, ZERO, ONE};
+    }
+    static constexpr vector<CFP_t> getToffoli() {
+        return {ONE,  ZERO, ZERO, ZERO, ZERO, ZERO, ZERO, ZERO, ZERO, ONE,
+                ZERO, ZERO, ZERO, ZERO, ZERO, ZERO, ZERO, ZERO, ONE,  ZERO,
+                ZERO, ZERO, ZERO, ZERO, ZERO, ZERO, ZERO, ONE,  ZERO, ZERO,
+                ZERO, ZERO, ZERO, ZERO, ZERO, ZERO, ONE,  ZERO, ZERO, ZERO,
+                ZERO, ZERO, ZERO, ZERO, ZERO, ONE,  ZERO, ZERO, ZERO, ZERO,
+                ZERO, ZERO, ZERO, ZERO, ZERO, ONE,  ZERO, ZERO, ZERO, ZERO,
+                ZERO, ZERO, ONE,  ZERO};
+    }
```
Yes. These were all defined for each gate class so I implemented them for brevity. Since we do not use them here though it may be fine to move them to a separate header/class. What do you think?

trbromley replied:
Could we just remove them entirely (unless we expect to use in future?)

mlxd replied:
Yep, pulling them out of here seems fine. I use some of these for the tests, and I think it will be good practice to allow pulling the gate definition out as needed, so will likely just move these to their own namespace `Pennylane::Gates` and allow them to be called from there.

trevor-vincent commented on pennylane_lightning/src/StateVector.hpp:
```diff
@@ -15,19 +15,764 @@
+    /**
+     * @brief Get indices not participating in operation.
+     *
+     * @param indicesToExclude
```
Missing some description here?

mlxd replied:
Yea, the docs need to follow. We have a story created for this, as the entire class docs need an update.

trevor-vincent reviewed (changes requested):
Really awesome work Lee. I left a few comments. Since I just joined, it might be due to my lack of familiarity, but I was wondering why there are gate definitions in two different places in the code, e.g. getCSWAP in StateVector.hpp and CSWAP in GateData.hpp, is this a necessity?

mlxd commented on CMakeLists.txt:
```diff
@@ -20,10 +20,9 @@ set_pennylane_lightning_version("${CMAKE_SOURCE_DIR}/pennylane_lightning/_versio
 
-set(CMAKE_CXX_STANDARD 11) # At least C++11 is required
+set(CMAKE_CXX_STANDARD 17) # At least C++17 is required
 find_package(OpenMP REQUIRED) # find OpenMP
```
GateData.hpp was a carry=over from the googletest suite. I updated them, then decided to pull them out. You are completely correct that it can be pulled. I'll removed all the remants of the old testing suite.

mlxd replied:
After a little thought, and since there are a few other places and libs to tackle the removal of, I'd rather move this to a separate story, and keep this PR to the templating.

mlxd commented on pennylane_lightning/src/StateVector.hpp:
```diff
@@ -15,19 +15,764 @@
+    const FMap gates_;
+    const std::unordered_map<string, size_t> gate_wires_;
+
+    CFP_t *const arr_;
```
This is a carry-over of the original design which does state-prep in numpy. To access the numpy data we point the above pointer to it via the available binding logic. I wish to completely remove the numpy prep (to be followed in a subsequent PR) and replace it with a std::vector in Cpp.

For now, I'd rather not tamper with the existing implementation (shared would be best, but not sure how numpy would handle it), since CPP doesn't own the pointer, merely just operates on it.

trevor-vincent replied:
Sounds like a good plan

mlxd commented:
> Really awesome work Lee. I left a few comments. Since I just joined, it might be due to my lack of familiarity, but I was wondering why there are gate definitions in two different places in the code, e.g. getCSWAP in StateVector.hpp and CSWAP in GateData.hpp, is this a necessity?

GateData.hpp was a carry-over from the googletest suite. I updated them, then decided to pull them out. You are completely correct that it can be pulled. I'll removed all the remnants of the old testing suite.

mlxd commented:
Recent changes have pulled the gate definitions from the class (fixes the CodeFactor complaint on StateVector), moved them to their own separate namespace, and used these to assist in the tests. This also fixed the second CodeFactor complaint. 

I think anything else needed removal of unused libs can be moved to a separate PR and story once we have some spare time.

trevor-vincent reviewed (approved):
Looks good!

trbromley reviewed (approved):
:100:

mlxd commented:
I will update the implementation for the new gate, then merge.
